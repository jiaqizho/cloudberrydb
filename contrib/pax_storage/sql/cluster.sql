set pax_max_tuples_per_file to 131072;

-- cluster table using index
-- start_ignore
drop table if EXISTS t_index_cluster;
-- end_ignore
create table t_index_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');

-- error, becuase no cluster index or cluster_columns is defined
-- cluster by btree index
create index on t_index_cluster(c1);
cluster t_index_cluster using t_index_cluster_c1_idx;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');

drop table t_index_cluster;

-- cluster table

-- zorder cluster
-- start_ignore
drop table if EXISTS t_zorder_cluster;
-- end_ignore
create table t_zorder_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
\d+ t_zorder_cluster;

insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');

-- error, becuase no cluster index or cluster_columns is defined
cluster t_zorder_cluster;

alter table t_zorder_cluster set(cluster_columns='c1,c2');
\d+ t_zorder_cluster;
-- success
cluster t_zorder_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');


drop table t_zorder_cluster;
-- test cluster reloption without order
create table t_zorder_cluster(c1 int, c2 int) with (minmax_columns='c1,c2', cluster_columns='c2,c1');
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
cluster t_zorder_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');

drop table t_zorder_cluster;

-- test cluster index
set pax_max_tuples_per_file to 131072;
drop table if EXISTS t_index_cluster;
create table t_index_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
\d+ t_index_cluster;

insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');

-- error, becuase no cluster index or cluster_columns is defined
cluster t_index_cluster;

create index idx_t_index_cluster on t_index_cluster(c1);
\d+ t_index_cluster;

-- success
cluster t_index_cluster using idx_t_index_cluster;

select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
drop table t_index_cluster;

-- test both cluster index and cluster columns
create table t_both_zorder_and_index_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
create index idx_t_both_zorder_and_index_cluster on t_both_zorder_and_index_cluster(c1);
-- error, because zorder cluster and index cluster can not be used together
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
-- success
cluster t_both_zorder_and_index_cluster;
alter table t_both_zorder_and_index_cluster set(cluster_columns='');
-- error, because zorder cluster reloptions has been deleted
cluster t_both_zorder_and_index_cluster;
-- success
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
-- error
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
drop index idx_t_both_zorder_and_index_cluster;
-- error
cluster t_both_zorder_and_index_cluster;
-- success
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
-- success
cluster t_both_zorder_and_index_cluster;

drop table t_both_zorder_and_index_cluster;

-- test unsupport type

create table t_zorder_unsupport_type(c1 int, c2 numeric(10,2),c3 varchar(128), c4 timestamp, c5 bpchar(64) ) with (minmax_columns='c1,c2');

-- error, because numeric is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c2');
alter table t_zorder_unsupport_type set(cluster_columns='c1,c3');
-- error, because timestamp is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c4');
alter table t_zorder_unsupport_type set(cluster_columns='c1,c5');

drop table t_zorder_unsupport_type;


-- lexical cluster
create table t_lexical_cluster(c1 text, c2 int) with (minmax_columns='c1,c2');
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_lexical_cluster');

-- set zorder cluster
alter table t_lexical_cluster set(cluster_columns='c1,c2');
cluster t_lexical_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_lexical_cluster');

-- set lexical cluster
alter table t_lexical_cluster set(cluster_columns='c1,c2', cluster_type='lexical');
cluster t_lexical_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_lexical_cluster');

drop table t_lexical_cluster;