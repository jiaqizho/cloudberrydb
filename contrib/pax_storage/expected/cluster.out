set pax_max_tuples_per_file to 131072;
-- cluster table using index
-- start_ignore
drop table if EXISTS t_index_cluster;
-- end_ignore
create table t_index_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
 ptblockname |                                          ptstatistics                                           | ptisclustered 
-------------+-------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           1 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           2 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           3 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           4 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           0 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           1 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           2 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           3 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           4 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           0 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           1 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           2 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           3 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           4 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
(15 rows)

-- error, becuase no cluster index or cluster_columns is defined
-- cluster by btree index
create index on t_index_cluster(c1);
cluster t_index_cluster using t_index_cluster_c1_idx;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
 ptblockname |                                              ptstatistics                                               | ptisclustered 
-------------+---------------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(131072),(5,79072),(5187913809)],[(false,false),(131072),(5,79072),(5187913809)]         | t
           1 | [(false,false),(34983),(79072,100000),(3132737841)],[(false,false),(34983),(79072,100000),(3132737841)] | t
           0 | [(false,false),(131072),(2,78230),(5147801190)],[(false,false),(131072),(2,78230),(5147801190)]         | t
           1 | [(false,false),(36238),(78230,99999),(3226202455)],[(false,false),(36238),(78230,99999),(3226202455)]   | t
           0 | [(false,false),(131072),(1,78633),(5126644321)],[(false,false),(131072),(1,78633),(5126644321)]         | t
           1 | [(false,false),(35563),(78633,99996),(3178950384)],[(false,false),(35563),(78633,99996),(3178950384)]   | t
(6 rows)

drop table t_index_cluster;
-- cluster table
-- zorder cluster
-- start_ignore
drop table if EXISTS t_zorder_cluster;
-- end_ignore
create table t_zorder_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
\d+ t_zorder_cluster;
                             Table "public.t_zorder_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Distributed by: (c1)
Options: minmax_columns=c1,c2

insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');
 ptblockname |                                          ptstatistics                                           | ptisclustered 
-------------+-------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           1 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           2 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           3 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           4 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           0 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           1 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           2 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           3 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           4 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           0 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           1 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           2 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           3 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           4 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
(15 rows)

-- error, becuase no cluster index or cluster_columns is defined
cluster t_zorder_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_zorder_cluster"
alter table t_zorder_cluster set(cluster_columns='c1,c2');
\d+ t_zorder_cluster;
                             Table "public.t_zorder_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Distributed by: (c1)
Options: minmax_columns=c1,c2, cluster_columns=c1,c2

-- success
cluster t_zorder_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');
 ptblockname |                                              ptstatistics                                               | ptisclustered 
-------------+---------------------------------------------------------------------------------------------------------+---------------
           5 | [(false,false),(131072),(2,78230),(5147801190)],[(false,false),(131072),(2,78230),(5147801190)]         | t
           6 | [(false,false),(36238),(78230,99999),(3226202455)],[(false,false),(36238),(78230,99999),(3226202455)]   | t
           5 | [(false,false),(131072),(1,78633),(5126644321)],[(false,false),(131072),(1,78633),(5126644321)]         | t
           6 | [(false,false),(35563),(78633,99996),(3178950384)],[(false,false),(35563),(78633,99996),(3178950384)]   | t
           5 | [(false,false),(131072),(5,79072),(5187913809)],[(false,false),(131072),(5,79072),(5187913809)]         | t
           6 | [(false,false),(34983),(79072,100000),(3132737841)],[(false,false),(34983),(79072,100000),(3132737841)] | t
(6 rows)

drop table t_zorder_cluster;
-- test cluster reloption without order
create table t_zorder_cluster(c1 int, c2 int) with (minmax_columns='c1,c2', cluster_columns='c2,c1');
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'c1' as the Cloudberry Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
insert into t_zorder_cluster select i,i from generate_series(1,100000) i;
cluster t_zorder_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_zorder_cluster');
 ptblockname |                                              ptstatistics                                               | ptisclustered 
-------------+---------------------------------------------------------------------------------------------------------+---------------
           5 | [(false,false),(131072),(1,78633),(5126644321)],[(false,false),(131072),(1,78633),(5126644321)]         | t
           6 | [(false,false),(35563),(78633,99996),(3178950384)],[(false,false),(35563),(78633,99996),(3178950384)]   | t
           5 | [(false,false),(131072),(5,79072),(5187913809)],[(false,false),(131072),(5,79072),(5187913809)]         | t
           6 | [(false,false),(34983),(79072,100000),(3132737841)],[(false,false),(34983),(79072,100000),(3132737841)] | t
           5 | [(false,false),(131072),(2,78230),(5147801190)],[(false,false),(131072),(2,78230),(5147801190)]         | t
           6 | [(false,false),(36238),(78230,99999),(3226202455)],[(false,false),(36238),(78230,99999),(3226202455)]   | t
(6 rows)

drop table t_zorder_cluster;
-- test cluster index
set pax_max_tuples_per_file to 131072;
drop table if EXISTS t_index_cluster;
create table t_index_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
\d+ t_index_cluster;
                              Table "public.t_index_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Distributed by: (c1)
Options: minmax_columns=c1,c2

insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
insert into t_index_cluster select i,i from generate_series(1,100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
 ptblockname |                                          ptstatistics                                           | ptisclustered 
-------------+-------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           1 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           2 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           3 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           4 | [(false,false),(33211),(5,100000),(1664130330)],[(false,false),(33211),(5,100000),(1664130330)] | f
           0 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           1 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           2 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           3 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           4 | [(false,false),(33327),(1,99996),(1661118941)],[(false,false),(33327),(1,99996),(1661118941)]   | f
           0 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           1 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           2 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           3 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
           4 | [(false,false),(33462),(2,99999),(1674800729)],[(false,false),(33462),(2,99999),(1674800729)]   | f
(15 rows)

-- error, becuase no cluster index or cluster_columns is defined
cluster t_index_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_index_cluster"
create index idx_t_index_cluster on t_index_cluster(c1);
\d+ t_index_cluster;
                              Table "public.t_index_cluster"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+---------+--------------+-------------
 c1     | integer |           |          |         | plain   |              | 
 c2     | integer |           |          |         | plain   |              | 
Indexes:
    "idx_t_index_cluster" btree (c1)
Distributed by: (c1)
Options: minmax_columns=c1,c2

-- success
cluster t_index_cluster using idx_t_index_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_index_cluster');
 ptblockname |                                              ptstatistics                                               | ptisclustered 
-------------+---------------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(131072),(2,78230),(5147801190)],[(false,false),(131072),(2,78230),(5147801190)]         | t
           1 | [(false,false),(36238),(78230,99999),(3226202455)],[(false,false),(36238),(78230,99999),(3226202455)]   | t
           0 | [(false,false),(131072),(1,78633),(5126644321)],[(false,false),(131072),(1,78633),(5126644321)]         | t
           1 | [(false,false),(35563),(78633,99996),(3178950384)],[(false,false),(35563),(78633,99996),(3178950384)]   | t
           0 | [(false,false),(131072),(5,79072),(5187913809)],[(false,false),(131072),(5,79072),(5187913809)]         | t
           1 | [(false,false),(34983),(79072,100000),(3132737841)],[(false,false),(34983),(79072,100000),(3132737841)] | t
(6 rows)

drop table t_index_cluster;
-- test both cluster index and cluster columns
create table t_both_zorder_and_index_cluster(c1 int, c2 int) with (minmax_columns='c1,c2');
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
create index idx_t_both_zorder_and_index_cluster on t_both_zorder_and_index_cluster(c1);
-- error, because zorder cluster and index cluster can not be used together
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
ERROR:  cannot using index to cluster table which has cluster-columns (pax_access_handle.cc:1310)
-- success
cluster t_both_zorder_and_index_cluster;
alter table t_both_zorder_and_index_cluster set(cluster_columns='');
-- error, because zorder cluster reloptions has been deleted
cluster t_both_zorder_and_index_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_both_zorder_and_index_cluster"
-- success
cluster t_both_zorder_and_index_cluster using idx_t_both_zorder_and_index_cluster;
-- error
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
ERROR:  pax table has previously clustered index, can't set cluster_columns reloptions (pax_access_handle.cc:1423)
drop index idx_t_both_zorder_and_index_cluster;
-- error
cluster t_both_zorder_and_index_cluster;
ERROR:  there is no previously clustered index or cluster_columns reloptions for table "t_both_zorder_and_index_cluster"
-- success
alter table t_both_zorder_and_index_cluster set(cluster_columns='c1,c2');
-- success
cluster t_both_zorder_and_index_cluster;
drop table t_both_zorder_and_index_cluster;
-- test unsupport type
create table t_zorder_unsupport_type(c1 int, c2 numeric(10,2),c3 varchar(128), c4 timestamp, c5 bpchar(64) ) with (minmax_columns='c1,c2');
-- error, because numeric is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c2');
ERROR:  the type of column c2 does not support zorder cluster (pax_access_handle.cc:1434)
alter table t_zorder_unsupport_type set(cluster_columns='c1,c3');
-- error, because timestamp is unsupport type
alter table t_zorder_unsupport_type set(cluster_columns='c1,c4');
ERROR:  the type of column c4 does not support zorder cluster (pax_access_handle.cc:1434)
alter table t_zorder_unsupport_type set(cluster_columns='c1,c5');
drop table t_zorder_unsupport_type;
-- lexical cluster
create table t_lexical_cluster(c1 text, c2 int) with (minmax_columns='c1,c2');
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
INSERT INTO t_lexical_cluster (c1, c2) SELECT CONCAT('xxxxxxxx', i::text), i FROM generate_series(1, 100000) i;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_lexical_cluster');
 ptblockname |                                                ptstatistics                                                | ptisclustered 
-------------+------------------------------------------------------------------------------------------------------------+---------------
           0 | [(false,false),(33221),(xxxxxxxx1,xxxxxxxx99997),None],[(false,false),(33221),(1,99997),(1662682392)]      | f
           1 | [(false,false),(33221),(xxxxxxxx1,xxxxxxxx99997),None],[(false,false),(33221),(1,99997),(1662682392)]      | f
           2 | [(false,false),(33221),(xxxxxxxx1,xxxxxxxx99997),None],[(false,false),(33221),(1,99997),(1662682392)]      | f
           3 | [(false,false),(33221),(xxxxxxxx1,xxxxxxxx99997),None],[(false,false),(33221),(1,99997),(1662682392)]      | f
           4 | [(false,false),(33221),(xxxxxxxx1,xxxxxxxx99997),None],[(false,false),(33221),(1,99997),(1662682392)]      | f
           0 | [(false,false),(33488),(xxxxxxxx10,xxxxxxxx99999),None],[(false,false),(33488),(2,100000),(1669962392)]    | f
           1 | [(false,false),(33488),(xxxxxxxx10,xxxxxxxx99999),None],[(false,false),(33488),(2,100000),(1669962392)]    | f
           2 | [(false,false),(33488),(xxxxxxxx10,xxxxxxxx99999),None],[(false,false),(33488),(2,100000),(1669962392)]    | f
           3 | [(false,false),(33488),(xxxxxxxx10,xxxxxxxx99999),None],[(false,false),(33488),(2,100000),(1669962392)]    | f
           4 | [(false,false),(33488),(xxxxxxxx10,xxxxxxxx99999),None],[(false,false),(33488),(2,100000),(1669962392)]    | f
           0 | [(false,false),(33291),(xxxxxxxx10001,xxxxxxxx99998),None],[(false,false),(33291),(11,99998),(1667405216)] | f
           1 | [(false,false),(33291),(xxxxxxxx10001,xxxxxxxx99998),None],[(false,false),(33291),(11,99998),(1667405216)] | f
           2 | [(false,false),(33291),(xxxxxxxx10001,xxxxxxxx99998),None],[(false,false),(33291),(11,99998),(1667405216)] | f
           3 | [(false,false),(33291),(xxxxxxxx10001,xxxxxxxx99998),None],[(false,false),(33291),(11,99998),(1667405216)] | f
           4 | [(false,false),(33291),(xxxxxxxx10001,xxxxxxxx99998),None],[(false,false),(33291),(11,99998),(1667405216)] | f
(15 rows)

-- set zorder cluster
alter table t_lexical_cluster set(cluster_columns='c1,c2');
cluster t_lexical_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_lexical_cluster');
 ptblockname |                                                  ptstatistics                                                   | ptisclustered 
-------------+-----------------------------------------------------------------------------------------------------------------+---------------
           5 | [(false,false),(131072),(xxxxxxxx10001,xxxxxxxx9998),None],[(false,false),(131072),(11,78956),(5169426427)]     | t
           6 | [(false,false),(35383),(xxxxxxxx78956,xxxxxxxx99998),None],[(false,false),(35383),(78956,99998),(3167599653)]   | t
           5 | [(false,false),(131072),(xxxxxxxx1,xxxxxxxx9999),None],[(false,false),(131072),(1,78615),(5184615315)]          | t
           6 | [(false,false),(35033),(xxxxxxxx78615,xxxxxxxx99997),None],[(false,false),(35033),(78615,99997),(3128796645)]   | t
           5 | [(false,false),(131072),(xxxxxxxx10,xxxxxxxx9996),None],[(false,false),(131072),(2,78366),(5108185712)]         | t
           6 | [(false,false),(36368),(xxxxxxxx100000,xxxxxxxx99999),None],[(false,false),(36368),(78366,100000),(3241626248)] | t
(6 rows)

-- set lexical cluster
alter table t_lexical_cluster set(cluster_columns='c1,c2', cluster_type='lexical');
cluster t_lexical_cluster;
select ptblockname,ptstatistics,ptisclustered from get_pax_aux_table('t_lexical_cluster');
 ptblockname |                                                 ptstatistics                                                 | ptisclustered 
-------------+--------------------------------------------------------------------------------------------------------------+---------------
           7 | [(false,false),(131072),(xxxxxxxx1,xxxxxxxx8058),None],[(false,false),(131072),(1,80578),(5425285866)]       | t
           8 | [(false,false),(35033),(xxxxxxxx8058,xxxxxxxx99997),None],[(false,false),(35033),(9,99997),(2888126094)]     | t
           7 | [(false,false),(131072),(xxxxxxxx10001,xxxxxxxx81150),None],[(false,false),(131072),(11,81150),(5430623690)] | t
           8 | [(false,false),(35383),(xxxxxxxx81150,xxxxxxxx99998),None],[(false,false),(35383),(82,99998),(2906402390)]   | t
           7 | [(false,false),(131072),(xxxxxxxx10,xxxxxxxx80572),None],[(false,false),(131072),(2,100000),(5366436984)]    | t
           8 | [(false,false),(36368),(xxxxxxxx80572,xxxxxxxx99999),None],[(false,false),(36368),(83,99999),(2983374976)]   | t
(6 rows)

drop table t_lexical_cluster;
